File: fraiseql-clojure/src/fraiseql/schema.clj
(ns fraiseql.schema
  (:require [cheshire.core :as json]
            [clojure.java.io :as io])
  (:import [java.nio.file Files Paths]))

;; Central registry for GraphQL type definitions
(def ^:private registry (atom {}))

;; MARK: - Scope Validation

(defn validate-scope
  "Validates scope format: action:resource
   Examples: read:user.email, admin:*, write:Post.*
   Rules:
   - Action: [a-zA-Z_][a-zA-Z0-9_]*
   - Resource: [a-zA-Z_][a-zA-Z0-9_.]*|*
   - Global wildcard: *"
  [scope]
  (cond
    (nil? scope) false
    (= "" scope) false
    (= "*" scope) true
    :else
    (let [parts (clojure.string/split scope #":" 2)]
      (if (= 2 (count parts))
        (let [[action resource] parts]
          (and (not (empty? action))
               (not (empty? resource))
               (re-matches #"[a-zA-Z_][a-zA-Z0-9_]*" action)
               (or (= "*" resource)
                   (re-matches #"[a-zA-Z_][a-zA-Z0-9_.]*" resource))))
        false))))

(defn validate-field-scopes
  "Validates scope fields in all fields of a type"
  [fields type-name]
  (doseq [[field-name field-config] fields]
    (let [has-scope (contains? field-config :requires_scope)
          has-scopes (contains? field-config :requires_scopes)]

      ;; Check for conflicting scope and scopes
      (when (and has-scope has-scopes)
        (throw (Exception. (str "Field \"" field-name "\" cannot have both requires_scope and requires_scopes"))))

      ;; Validate requires_scope if present
      (when has-scope
        (let [scope (:requires_scope field-config)]
          (if (not (string? scope))
            (throw (Exception. (str "Field \"" field-name "\" requires_scope must be a string"))))
          (if (not (validate-scope scope))
            (throw (Exception. (str "Field \"" field-name "\" has invalid scope format: \"" scope "\""))))))

      ;; Validate requires_scopes if present
      (when has-scopes
        (let [scopes (:requires_scopes field-config)]
          (if (not (vector? scopes))
            (throw (Exception. (str "Field \"" field-name "\" requires_scopes must be a vector"))))
          (if (empty? scopes)
            (throw (Exception. (str "Field \"" field-name "\" requires_scopes cannot be empty"))))
          (doseq [scope scopes]
            (if (not (string? scope))
              (throw (Exception. (str "Field \"" field-name "\" requires_scopes contains non-string value"))))
            (if (not (validate-scope scope))
              (throw (Exception. (str "Field \"" field-name "\" has invalid scope in requires_scopes: \"" scope "\"")))))))))))

(defn register
  "Register a type definition in the schema registry"
  [name type-info]
  (swap! registry assoc name type-info))

(defn get-type-names
  "Get all registered type names"
  []
  (keys @registry))

(defn get-type
  "Get a registered type by name"
  [name]
  (get @registry name))

(defn clear
  "Clear all registered types (useful for testing)"
  []
  (reset! registry {}))

(defn register-type
  "Register a type definition

   name - The type name
   fields - Map of field name to field definition
   description - Optional type description"
  ([name fields]
   (register-type name fields nil))
  ([name fields description]
   ;; Validate scope fields in all fields
   (validate-field-scopes fields name)

   (register name {:name name
                   :fields fields
                   :description description})))

(defn export-types
  "Export minimal schema with only types (TOML workflow)

   Returns JSON containing only the \"types\" section.
   All operational configuration (queries, mutations, federation, security, observers)
   comes from fraiseql.toml and is merged during compilation.

   pretty - Pretty-print JSON (true = formatted, false = compact)"
  ([]
   (export-types true))
  ([pretty]
   (let [type-names (get-type-names)
         types (map (fn [type-name]
                      (when-let [type-info (get-type type-name)]
                        (let [fields (:fields type-info)
                              fields-array (map (fn [[field-name field-def]]
                                                  (let [field {:name field-name
                                                               :type (get field-def :type "String")
                                                               :nullable (get field-def :nullable false)}]
                                                    (cond-> field
                                                      (contains? field-def :requires_scope)
                                                      (assoc :requires_scope (:requires_scope field-def))
                                                      (contains? field-def :requires_scopes)
                                                      (assoc :requires_scopes (:requires_scopes field-def)))))
                                                fields)]
                          (cond-> {:name type-name
                                   :fields (vec fields-array)}
                            (:description type-info)
                            (assoc :description (:description type-info))))))
                    type-names)
         types (vec (filter identity types))
         minimal-schema {:types types}]
     (if pretty
       (json/generate-string minimal-schema {:pretty true})
       (json/generate-string minimal-schema)))))

(defn export-types-file
  "Export minimal types to a file

   output-path - File path for types.json"
  [output-path]
  (try
    (let [types-json (export-types true)
          path (Paths/get output-path (into-array String []))
          parent (.getParent path)]

      ;; Ensure directory exists
      (when parent
        (Files/createDirectories parent (into-array java.nio.file.attribute.FileAttribute [])))

      ;; Write file
      (spit output-path types-json)

      ;; Print summary
      (let [types-count (count (get-type-names))]
        (println (str "✅ Types exported to " output-path))
        (println (str "   Types: " types-count))
        (println)
        (println "🎯 Next steps:")
        (println (str "   1. fraiseql compile fraiseql.toml --types " output-path))
        (println "   2. This merges types with TOML configuration")
        (println "   3. Result: schema.compiled.json with types + all config")))
    (catch Exception e
      (throw (RuntimeException. (str "Failed to write types file: " output-path))))))

(defn reset
  "Reset schema registry (useful for testing)"
  []
  (clear))

File: sdks/official/fraiseql-rust/src/schema.rs
//! Schema registry and scope validation for field-level RBAC.
//!
//! Provides:
//! - `SchemaRegistry` for tracking types and their field scopes
//! - `validate_scope()` for validating scope format (action:resource)
//! - JSON export of schema with scope metadata

use crate::field::Field;
use serde::Serialize;
use std::collections::HashMap;

/// Error type for scope validation failures.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ScopeValidationError {
    /// Missing colon separator (format must be "action:resource")
    MissingColon,
    /// Empty action part (before colon)
    EmptyAction,
    /// Empty resource part (after colon)
    EmptyResource,
    /// Invalid action format (must be alphanumeric + underscore)
    InvalidAction(String),
    /// Invalid resource format (must be alphanumeric + underscore + dot, or wildcard)
    InvalidResource(String),
    /// Empty scope string
    EmptyScope,
}

/// Validates scope format: action:resource
///
/// # Valid formats
/// - `read:user.email` (specific field)
/// - `read:User.*` (all User fields)
/// - `admin:*` (global admin)
/// - `*` (wildcard)
///
/// # Invalid formats
/// - `readuser` (missing colon)
/// - `read-all:user` (hyphen in action)
/// - `read:user-data` (hyphen in resource)
/// - `:resource` (empty action)
/// - `action:` (empty resource)
///
/// # Errors
/// Returns `ScopeValidationError` for invalid formats.
pub fn validate_scope(scope: &str) -> Result<(), ScopeValidationError> {
    if scope.is_empty() {
        return Err(ScopeValidationError::EmptyScope);
    }

    // Global wildcard is always valid
    if scope == "*" {
        return Ok(());
    }

    // Check for colon separator
    if !scope.contains(':') {
        return Err(ScopeValidationError::MissingColon);
    }

    let parts: Vec<&str> = scope.splitn(2, ':').collect();
    if parts.len() != 2 {
        return Err(ScopeValidationError::MissingColon);
    }

    let action = parts[0];
    let resource = parts[1];

    // Check for empty parts
    if action.is_empty() {
        return Err(ScopeValidationError::EmptyAction);
    }
    if resource.is_empty() {
        return Err(ScopeValidationError::EmptyResource);
    }

    // Validate action format: [a-zA-Z_][a-zA-Z0-9_]*
    if !is_valid_action(action) {
        return Err(ScopeValidationError::InvalidAction(action.to_string()));
    }

    // Validate resource format: [a-zA-Z_][a-zA-Z0-9_.]*|*
    if !is_valid_resource(resource) {
        return Err(ScopeValidationError::InvalidResource(resource.to_string()));
    }

    Ok(())
}

/// Validates action format: [a-zA-Z_][a-zA-Z0-9_]*
fn is_valid_action(action: &str) -> bool {
    if action.is_empty() {
        return false;
    }

    let first_char = action.chars().next().unwrap();
    if !first_char.is_alphabetic() && first_char != '_' {
        return false;
    }

    action.chars().all(|c| c.is_alphanumeric() || c == '_')
}

/// Validates resource format: [a-zA-Z_][a-zA-Z0-9_.]*|*
fn is_valid_resource(resource: &str) -> bool {
    if resource == "*" {
        return true;
    }

    if resource.is_empty() {
        return false;
    }

    let first_char = resource.chars().next().unwrap();
    // First character must be letter (upper/lower) or underscore
    if !first_char.is_alphabetic() && first_char != '_' {
        return false;
    }

    // Remaining characters can be alphanumeric, underscore, dot, or asterisk
    resource
        .chars()
        .all(|c| c.is_alphanumeric() || c == '_' || c == '.' || c == '*')
}

/// Registry for GraphQL schema types and their field scope requirements.
///
/// Tracks all types and fields, extracts scope metadata, and exports
/// to JSON format for the compiler.
#[derive(Debug, Clone)]
pub struct SchemaRegistry {
    types: HashMap<String, RegisteredType>,
}

/// One registered type: its fields and the SQL relation backing it.
#[derive(Debug, Clone)]
struct RegisteredType {
    fields: Vec<Field>,
    sql_source: Option<String>,
}

/// A type as it is written to `schema.json`.
#[derive(Serialize)]
struct SerializedType<'a> {
    name: &'a str,
    #[serde(skip_serializing_if = "Option::is_none")]
    sql_source: Option<&'a str>,
    fields: Vec<Field>,
}

/// The document written to `schema.json`.
///
/// The previous export produced a bare map keyed by type name — `{"User": {...}}` — with
/// no `types` key at all. The compiler reads `types` as an array of objects, so it saw a
/// schema with **zero** types, reported `✓ Schema compiled successfully`, and wrote an
/// artifact containing none of the author's work (#855). A silent success is the one
/// outcome worse than a rejection.
#[derive(Serialize)]
struct SerializedSchema<'a> {
    version: &'a str,
    types: Vec<SerializedType<'a>>,
}

impl SchemaRegistry {
    /// Creates a new empty schema registry.
    #[must_use]
    pub fn new() -> Self {
        Self {
            types: HashMap::new(),
        }
    }

    /// Registers a new type with its fields.
    ///
    /// # Arguments
    /// * `type_name` - Name of the type (e.g., "User", "Post")
    /// * `fields` - Vector of Field definitions
    pub fn register_type(&mut self, type_name: &str, fields: Vec<Field>) {
        self.types.insert(
            type_name.to_string(),
            RegisteredType {
                fields,
                sql_source: None,
            },
        );
    }

    /// Registers a type together with the SQL view backing it.
    ///
    /// # Arguments
    /// * `type_name` - Name of the type (e.g., "User", "Post")
    /// * `fields` - Vector of Field definitions
    /// * `sql_source` - The backing view, e.g. `v_user`
    pub fn register_type_with_source(
        &mut self,
        type_name: &str,
        fields: Vec<Field>,
        sql_source: &str,
    ) {
        self.types.insert(
            type_name.to_string(),
            RegisteredType {
                fields,
                sql_source: Some(sql_source.to_string()),
            },
        );
    }

    /// Gets registered fields for a type.
    ///
    /// # Arguments
    /// * `type_name` - Name of the type
    ///
    /// # Returns
    /// Option containing reference to field vector, or None if type not found
    #[must_use]
    pub fn get_type(&self, type_name: &str) -> Option<&Vec<Field>> {
        self.types.get(type_name).map(|registered| &registered.fields)
    }

    /// Extracts all fields that have scope requirements.
    ///
    /// # Returns
    /// `HashMap` mapping type name to vector of field names with scopes
    ///
    /// # Example
    /// ```
    /// // Returns: { "User": ["email", "password"] }
    /// ```
    #[must_use]
    pub fn extract_scoped_fields(&self) -> HashMap<String, Vec<String>> {
        let mut scoped = HashMap::new();

        for (type_name, registered) in &self.types {
            let scoped_fields: Vec<String> = registered
                .fields
                .iter()
                .filter(|f| f.requires_scope.is_some() || f.requires_scopes.is_some())
                .map(|f| f.name.clone())
                .collect();

            if !scoped_fields.is_empty() {
                scoped.insert(type_name.clone(), scoped_fields);
            }
        }

        scoped
    }

    /// Exports the registry as a `schema.json` document for `fraiseql compile`.
    ///
    /// # Returns
    /// A JSON string of the form `{"version": "2.0.0", "types": [...]}`.
    ///
    /// Built with `serde_json`, not string concatenation: the previous implementation
    /// interpolated raw values into JSON string literals, so a single `"` in a name,
    /// scope or description produced output that is not parseable JSON (#855).
    ///
    /// # Panics
    ///
    /// Panics when a field declares more than one required scope; see
    /// [`Field::normalized`].
    #[must_use]
    pub fn export_to_json(&self) -> String {
        // Sorted so the export is byte-stable across runs — `HashMap` iteration order is
        // not, and an unstable artifact makes every diff of a checked-in schema.json noise.
        let mut names: Vec<&String> = self.types.keys().collect();
        names.sort();

        let document = SerializedSchema {
            version: "2.0.0",
            types: names
                .into_iter()
                .map(|name| {
                    let registered = &self.types[name];
                    SerializedType {
                        name,
                        sql_source: registered.sql_source.as_deref(),
                        fields: registered.fields.iter().map(Field::normalized).collect(),
                    }
                })
                .collect(),
        };

        serde_json::to_string_pretty(&document)
            .expect("the schema document contains only strings and bools")
    }
}

impl Default for SchemaRegistry {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_validate_scope_valid_specific_field() {
        assert!(validate_scope("read:user.email").is_ok());
    }

    #[test]
    fn test_validate_scope_valid_wildcard_resource() {
        assert!(validate_scope("read:User.*").is_ok());
    }

    #[test]
    fn test_validate_scope_valid_global_wildcard() {
        assert!(validate_scope("admin:*").is_ok());
    }

    #[test]
    fn test_validate_scope_valid_pure_wildcard() {
        assert!(validate_scope("*").is_ok());
    }

    #[test]
    fn test_validate_scope_invalid_missing_colon() {
        assert!(validate_scope("readuser").is_err());
    }

    #[test]
    fn test_validate_scope_invalid_empty_string() {
        assert!(validate_scope("").is_err());
    }

    #[test]
    fn test_validate_scope_invalid_action_with_hyphen() {
        assert!(validate_scope("read-all:user").is_err());
    }

    #[test]
    fn test_validate_scope_invalid_resource_with_hyphen() {
        assert!(validate_scope("read:user-data").is_err());
    }

    #[test]
    fn test_schema_registry_new() {
        let registry = SchemaRegistry::new();
        assert_eq!(registry.types.len(), 0);
    }

    #[test]
    fn test_schema_registry_register_type() {
        let mut registry = SchemaRegistry::new();
        let fields = vec![Field::new("id", "Int")];
        registry.register_type("User", fields);

        assert!(registry.get_type("User").is_some());
        assert_eq!(registry.get_type("User").unwrap().len(), 1);
    }

    #[test]
    fn test_schema_registry_extract_scoped_fields() {
        let mut registry = SchemaRegistry::new();
        let fields = vec![
            Field::new("id", "Int"),
            Field::new("email", "String").with_requires_scope(Some("read:user.email".to_string())),
        ];
        registry.register_type("User", fields);

        let scoped = registry.extract_scoped_fields();
        assert!(scoped.contains_key("User"));
        assert_eq!(scoped["User"].len(), 1);
        assert_eq!(scoped["User"][0], "email");
    }
}
